File: src-tauri/src/query/compile.rs
use super::validate::ValidatedQuery;
use crate::query::ir::{BindValue, CompiledQuery, QuerySpec};
use crate::types::Dialect;

#[path = "compile_helpers.rs"]
mod compile_helpers;

#[cfg(any(test, feature = "test-helpers"))]
pub fn compile(spec: &QuerySpec, dialect: Dialect) -> Result<CompiledQuery, String> {
    compile_spec(spec, dialect, None)
}

pub fn compile_validated(
    validated: &ValidatedQuery,
    dialect: Dialect,
) -> Result<CompiledQuery, String> {
    compile_spec(validated.spec(), dialect, Some(validated.columns()))
}

fn compile_spec(
    spec: &QuerySpec,
    dialect: Dialect,
    validated_columns: Option<&[super::validate::ValidatedColumn]>,
) -> Result<CompiledQuery, String> {
    let mut params: Vec<BindValue> = Vec::new();
    let mut param_idx = 1u32;

    let select_clause = compile_helpers::build_select_clause(spec, dialect, validated_columns);
    let from_clause = compile_helpers::build_from_clause(spec, dialect);
    let join_clause = compile_helpers::build_join_clause(spec, dialect);
    let where_clause = compile_helpers::build_where_root(
        &spec.filters,
        &spec.connector_overrides,
        dialect,
        &mut params,
        &mut param_idx,
    )?;

    let mut sql = String::new();
    sql.push_str("SELECT ");

    let fetch_limit = spec.limit.saturating_add(1);

    if dialect == Dialect::Mssql {
        sql.push_str(&format!("TOP {} ", fetch_limit));
    }

    sql.push_str(&select_clause);
    sql.push('\n');
    sql.push_str("FROM ");
    sql.push_str(&from_clause);
    if !join_clause.is_empty() {
        sql.push('\n');
        sql.push_str(&join_clause);
    }
    if !where_clause.is_empty() {
        sql.push('\n');
        sql.push_str("WHERE ");
        sql.push_str(&where_clause);
    }

    match dialect {
        Dialect::Mssql => {}
        Dialect::Oracle => {
            sql.push_str("\nFETCH FIRST ");
            sql.push_str(&fetch_limit.to_string());
            sql.push_str(" ROWS ONLY");
        }
        _ => {
            sql.push_str("\nLIMIT ");
            sql.push_str(&fetch_limit.to_string());
        }
    }

    Ok(CompiledQuery { sql, params })
}

File: shared/src/main/kotlin/com/safedb/query/Compile.kt
package com.safedb.query

import com.safedb.model.BindValue
import com.safedb.model.CompiledQuery
import com.safedb.model.Dialect
import com.safedb.model.Outcome

fun compileValidated(validated: ValidatedQuery, dialect: Dialect): Outcome<CompiledQuery> {
    val spec = validated.spec()
    val params = mutableListOf<BindValue>()

    val selectClause = buildSelectClause(validated.columns(), dialect)
    val fromClause = buildFromClause(spec, dialect)
    val joinClause = buildJoinClause(spec, dialect)
    val whereClause =
        when (
            val result =
                buildWhereRoot(
                    group = spec.filters,
                    overrides = spec.connectorOverrides,
                    dialect = dialect,
                    params = params,
                )
        ) {
            is Outcome.Ok -> result.value
            is Outcome.Err -> return Outcome.err(result.message)
        }
    val orderByClause = buildOrderByClause(spec, dialect)
    val groupByClause = buildGroupByClause(spec, dialect)

    val fetchLimit = spec.limit + 1
    val sql = buildString {
        append("SELECT ")
        if (spec.distinct) {
            append("DISTINCT ")
        }
        if (dialect == Dialect.Mssql) {
            append("TOP $fetchLimit ")
        }
        append(selectClause)
        append('\n')
        append("FROM ")
        append(fromClause)
        if (joinClause.isNotEmpty()) {
            append('\n')
            append(joinClause)
        }
        if (whereClause.isNotEmpty()) {
            append('\n')
            append("WHERE ")
            append(whereClause)
        }
        if (groupByClause.isNotEmpty()) {
            append("\nGROUP BY ")
            append(groupByClause)
        }
        if (orderByClause.isNotEmpty()) {
            append("\nORDER BY ")
            append(orderByClause)
        }
        when (dialect) {
            Dialect.Mssql -> Unit
            Dialect.Oracle -> {
                append("\nFETCH FIRST $fetchLimit ROWS ONLY")
            }
            else -> {
                append("\nLIMIT $fetchLimit")
            }
        }
    }

    return Outcome.ok(CompiledQuery(sql = sql, params = params.toList()))
}
